feat(vdo): Add safe Rust bindings for VDO API

vsem-azamat opened the pull request:
## Safe Rust bindings for VDO (Video Capture) API

Safe wrappers over `vdo-sys` with builder pattern for stream creation, iterator-based frame capture, and automatic resource cleanup.

Tested on device: YUV, JPEG, H.264, H.265 formats work correctly.

### Describe your changes

This PR adds a new `vdo` crate providing safe Rust bindings for the VDO API, enabling video capture from Axis cameras.

**Based on the work from PR #153 by @JsGjKJzi** — completed the implementation, added comprehensive tests, and updated the example application.
> Note: @JsGjKJzi hasn't responded since January 2024.

Key features:
- `StreamBuilder` with builder pattern for configuring video streams
- `Stream`, `RunningStream`, `StreamIterator` for lifecycle management
- `StreamBuffer` and `Frame` for zero-copy frame access
- Automatic resource cleanup via `Drop` implementations
- Comprehensive error handling with `VdoError` including error code names
- Platform compatibility documentation (Artpec-6/7/8/9, Ambarella CV)

Changes:
- New `crates/vdo/` crate with safe bindings
- Updated `apps/vdo_encode_client/` to use the new `vdo` crate
- Unit tests (no device required) + device tests (gated by `device-tests` feature)

### Issue ticket number and link

- Continues #153
- Enables #120 (Larod ML API depends on VDO)

### Checklist before requesting a review

- [x] I have performed a self-review of my own code
- [x] I have verified that the code builds perfectly fine on my local system
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have verified that my code follows the style already available in the repository
- [x] I have made corresponding changes to the documentation

vsem-azamat committed:
feat(vdo): Add safe Rust bindings for VDO API

Safe wrappers over vdo-sys with builder pattern for stream creation,
iterator-based frame capture, and automatic resource cleanup.

Tested on device: YUV, JPEG, H.264, H.265 formats work correctly.

vsem-azamat commented:
I'll wait for feedback from the maintainers before addressing these suggestions, as some may be more relevant than others for this codebase.

apljungquist commented on crates/vdo/src/lib.rs:
```diff
@@ -0,0 +1,1359 @@
+//!
+//! # Known Issues
+//!
+//! - Image rotation may vary between platforms. Check the `rotation` property in stream info.
```
I like that you are documenting pitfalls with the underlying API that cannot easily be fixed in our bindings.

apljungquist commented on crates/vdo/src/lib.rs:
```diff
@@ -0,0 +1,1359 @@
+
+    /// Gets a string value, returning an owned copy.
+    ///
+    /// Returns `None` if the key doesn't exist or the value is null.
```
This will also return `None` if the value is not valid UTF-8, will it not?

https://doc.rust-lang.org/std/ffi/struct.CStr.html#method.to_str

apljungquist commented on crates/vdo/src/lib.rs:
```diff
@@ -0,0 +1,1359 @@
+
+impl Stream {
+    /// Creates a new stream builder.
+    pub fn builder() -> StreamBuilder {
```
Nice

apljungquist commented on crates/vdo/src/lib.rs:
```diff
@@ -0,0 +1,1359 @@
+impl Drop for Stream {
+    fn drop(&mut self) {
+        unsafe {
+            vdo_stream_stop(self.raw);
```
What happens if the stream is not running?
And why do we not need to decrease the ref count?

https://developer.axis.com/acap/api/src/api/vdostream/html/vdo-stream_8h.html#a36fa0021eb58d482c494163db9b22b61

apljungquist commented on crates/vdo/src/lib.rs:
```diff
@@ -0,0 +1,1359 @@
+///
+/// Created by calling [`Stream::start()`]. Use [`iter()`](RunningStream::iter)
+/// to get an iterator over frames.
+pub struct RunningStream<'a> {
```
Nice, I like differentiating between states with distinct types

apljungquist commented on crates/vdo/src/lib.rs:
```diff
@@ -0,0 +1,1359 @@
+// Device Tests (require actual Axis camera)
+// ============================================================================
+
+#[cfg(all(test, target_arch = "aarch64", feature = "device-tests"))]
```
Elsewhere we are using `#[cfg(not(any(target_arch = "x86_64", target_os = "macos")))]`. This is far from perfect and you are welcome to submit a separate PR that changes it (I too think that a feature is probably the right direction), but I think we should be consistent.

apljungquist commented on crates/vdo/src/lib.rs:
```diff
@@ -0,0 +1,1359 @@
+    }
+
+    #[test]
+    fn stream_starts_and_stops() -> std::result::Result<(), Box<dyn std::error::Error>> {
```
In my experience the error messages from failing tests are easier to understand when using `unwrap` instead of returning an error from the test function as a whole. One problem with returning an error is that we don't learn which line of the test failed, so the first step of troubleshooting is often adding this information. Super annoying when the failure is not deterministic/reproducible.

apljungquist commented on crates/vdo/src/lib.rs:
```diff
@@ -0,0 +1,1359 @@
+            // Verify JPEG magic bytes (SOI marker)
+            let data = buffer.as_slice()?;
+            assert!(data.len() >= 2, "Buffer too small for JPEG");
+            assert_eq!(data[0], 0xFF, "Invalid JPEG SOI marker");
```
Nice that we have tests actually inspecting the image data.

apljungquist commented on Cargo.toml:
```diff
@@ -68,6 +68,7 @@ licensekey-sys = { path = "crates/licensekey-sys" }
 mdb = { path = "crates/mdb" }
 mdb-sys = { path = "crates/mdb-sys" }
 vdo-sys = { path = "crates/vdo-sys" }
+vdo = { path = "crates/vdo" }
```
`vdo` should come before `vdo-sys`

apljungquist commented on crates/vdo/src/lib.rs:
```diff
@@ -0,0 +1,1359 @@
+            assert!(frame.size() <= capacity, "Frame size should be <= capacity");
+
+            // Verify we can read data without crashing
+            let _first_byte = data[0];
```
I like the idea and I think we should keep it. Maybe even add a variant for `as_mut_slice`. But I will still challenge you a little bit:

Are you sure these accesses aren't removed by the optimizer?
I seem to remember this being a pitfall when benchmarking: https://doc.rust-lang.org/stable/std/hint/fn.black_box.html

Also, if they are not removed, would they always fail if we cannot access the data or only if we are (un)lucky and the address points to a certain region of memory?

I think it may be worth documenting our answers to these questions, either with comments in the code or in the PR description which will later turn into the commit message.

apljungquist commented on crates/vdo/src/lib.rs:
```diff
@@ -0,0 +1,1359 @@
+    }
+
+    #[test]
+    fn multiple_streams_sequential() -> std::result::Result<(), Box<dyn std::error::Error>> {
```
Nice test. Have you considered the use case of concurrent streams?

apljungquist commented on crates/vdo/src/lib.rs:
```diff
@@ -0,0 +1,1359 @@
+    }
+
+    #[test]
+    fn error_message_is_descriptive() {
```
How does this test (automatically) verify that the error message is descriptive?

apljungquist reviewed (commented):
I skimmed the PR and I think it looks pretty good. I'll try to do a more thorough read in the before next weekend focusing on safety and I'll try to respond to comments within 24h.

apljungquist commented on crates/vdo/src/lib.rs:
```diff
@@ -0,0 +1,1359 @@
+
+    /// Sets a 32-bit unsigned integer value.
+    pub fn set_u32(&self, key: &str, value: u32) -> Result<()> {
+        let key_cstr = CString::new(key).map_err(|_| Error::CStringAllocation)?;
```
What do you think about taking the key as `&CStr`?

That would make this function infallible, relieving users of error handling.
In some cases creating the `CStr` would be annoying for the user, but in other cases they will be using `c"string"` literals anyway. It also gives the user the flexibility to decide how to handle conversion (errors), maybe they would rather do a lossy conversion.

apljungquist commented on crates/vdo/src/lib.rs:
```diff
@@ -0,0 +1,1359 @@
+    /// Sets the video resolution.
+    ///
+    /// If width or height is 0, the camera's native resolution is used.
+    pub fn resolution(mut self, width: u32, height: u32) -> Self {
```
What do you think about making width and height separate setter functions?
That way there is not ambiguity about which argument comes first.

It also makes it possible to make `0` an error. Or maybe we could have a `Resolution` enum with variants for `Native` and `Exact` (if that matches how the API works).

apljungquist commented on crates/vdo/src/lib.rs:
```diff
@@ -0,0 +1,1359 @@
+    /// Stops the stream.
+    ///
+    /// After stopping, no more frames can be retrieved.
+    pub fn stop(&mut self) -> Result<()> {
```
What do you think about taking `self` instead of `&mut self`?

That way it will be clear when a stream is stopped (no `.stop` which don't do anything) and it will definitely be impossible to `iter` on a stopped stream.

apljungquist commented on crates/vdo/src/lib.rs:
```diff
@@ -0,0 +1,1359 @@
+impl Drop for StreamBuffer<'_> {
+    fn drop(&mut self) {
+        unsafe {
+            let _ = try_func!(vdo_stream_buffer_unref, self.stream.raw, &mut self.raw);
```
What do you think about adding an `unref` method to `StreamBuffer` that allows the user to obtain the error message?

apljungquist commented on crates/vdo/src/lib.rs:
```diff
@@ -0,0 +1,1359 @@
+    }
+
+    /// Returns frame metadata for this buffer.
+    pub fn frame(&self) -> Result<Frame<'_>> {
```
What do you think about returning an `Option` instead of a `Result`?

This maps quite nicely to a nullable pointer and users don't have to think about if/how to handle all the other error variants.

apljungquist reviewed (commented):
Some API design related questions.

API design is difficult and typically subjective. I don't want it to stall this PR, so consider this inspiration. (that's not to say that the previous review was not subjective or disputable, only that this is even more so)

guoxe commented on crates/vdo/src/lib.rs:
```diff
@@ -0,0 +1,1359 @@
+/// Metadata for a video frame.
+///
+/// Contains information about frame timing, size, and type.
+pub struct Frame<'a> {
```
Consider merging the `Frame` and `StreamBuffer` structs since their lifetimes are very much linked.
As stated in https://developer.axis.com/acap/api/src/api/vdostream/html/vdo-buffer_8h.html#a5367ff8edb99976f0b22a2809d44c097, the frame object is only valid while the buffer is so I'm not convinced it needs a separate type when we could just add relevant methods to the StreamBuffer type.

apljungquist replied:
I'm not saying we should depend on this, but no shit they are linked: `typedef VdoFrame VdoBuffer;` :sweat_smile:

vsem-azamat replied:
`Frame` was merged into [`StreamBuffer`](https://github.com/vsem-azamat/acap-rs/blob/feature/vdo-completion/crates/vdo/src/lib.rs#L438). 
All frame metadata methods are now on StreamBuffer directly.

apljungquist commented on crates/vdo/src/lib.rs:
```diff
@@ -0,0 +1,1359 @@
+}
+
+/// Result type for VDO operations.
+pub type Result<T> = std::result::Result<T, Error>;
```
nit; I would prefer it if we don't alias `Result` because writing out `Result<Error>` isn't that much work and if we have multiple result types it starts getting difficult understanding what a function returns by just reading the signature.

apljungquist reviewed (commented):
I have focused this review on the `Map` type, which looks good. Once we have decided if/how to change it I will continue reviewing other parts.

I do this to not overwhelm you with comments and to not overwhelm myself with different discussions to keep track of and large diffs that need to be reviewed with every now commit.

vsem-azamat committed:
fix(vdo): Address all PR #223 review feedback

Address all review comments. Fix GObject leak in Stream::drop.

- Consume-by-value ownership: start(self), stop(self), unref(self)
- Replace Iterator with next_buffer() -> Result
- Merge Frame into StreamBuffer (VdoBuffer = VdoFrame)
- Extract Map to separate module, keys as &CStr
- Add Resolution enum, remove buffer_strategy, rename to custom_timestamp_us
- Replace as_mut_slice with data_copy(), file_descriptor() returns Result

Tested: cargo clippy clean, 33/33 tests pass on Artpec-9 device.

apljungquist commented:
Will try to get around to this tomorrow

apljungquist committed:
Merge remote-tracking branch 'upstream/main' into HEAD

apljungquist commented on crates/vdo/src/lib.rs:
```diff
@@ -1315,45 +1245,69 @@ mod device_tests {
     fn error_message_is_descriptive() {
```
For tests that verify that an output is subjectively good for a human, I have found that `expect-test` is a useful tool. You don't have to, but if you feel like looking into it I think it could be useful in this test.

apljungquist commented:
@vsem-azamat nice progress!

Please refrain from force pushing in the future as it makes it difficult to review the changes and keep track of comments. I took the liberty of cherry-picking your fix onto the branch before it was force pushed and force pushing that, so now the git history looks as it should.

A few of my comments seem to have gone unaddressed. You don't have to always to things the way I think is the best, but I would appreciate if you at least acknowledge on each comment that you have read it and state what, if anything, you will do to address it :slightly_smiling_face: 

I haven't seen the results of the CI yet, but if we get more revisions in on this I will make sure to start it as soon as I can instead of waiting until I have time to do human review like this time around.

apljungquist reviewed (commented):
I forgot to submit this review, sorry. Among the new comments I only feel strongly about the glob import, but as I said, please respond to all new and old comments.

vsem-azamat committed:
refactor(vdo): Replace glob imports with qualified paths and add expect-test

Replace `use vdo_sys::*` with explicit type imports and `vdo_sys::`-qualified
function/constant calls at call sites. This addresses the PR review feedback
about avoiding glob imports. Also adds the missing VDO_ERROR_NO_VIDEO error
code and adopts expect-test for display/formatting tests.

vsem-azamat committed:
fix(vdo): Address code review and security audit findings

- Clamp size to capacity in data_copy to prevent out-of-bounds reads
- Replace debug_assert with assert in unsafe constructors (from_ptr, from_raw)
- Document BorrowedFd lifetime constraints on file_descriptor()
- Fix double vdo_stream_stop by delegating to Drop
- Add Error::InvalidFd instead of overloading NullPointer for bad fds
- Return Option<usize> from header_size() for negative (no header) values
- Fix from_gerror safety comment to match actual code order
- Add Debug for Map and CStringPtr, Deref<Target=CStr> for CStringPtr
- Use explicit ptr::null::<c_char>(), comment as_ptr GLib convention

vsem-azamat committed:
style(vdo): Apply rustfmt

apljungquist commented on crates/vdo/src/map.rs:
```diff
@@ -18,7 +20,7 @@ impl CStringPtr {
     pub(crate) unsafe fn from_ptr(ptr: *mut c_char) -> Self {
-        debug_assert!(!ptr.is_null());
+        assert!(!ptr.is_null(), "CStringPtr::from_ptr called with null");
         Self(NonNull::new_unchecked(ptr))
```
Why not use `NonNull::new`?

https://doc.rust-lang.org/beta/src/core/ptr/non_null.rs.html#270

vsem-azamat replied:
I'll switch to `NonNull::new(...).expect(...)`.

apljungquist commented on crates/vdo/src/map.rs:
```diff
@@ -62,26 +78,27 @@ impl Map {
     /// `ptr` must be a non-null, valid `VdoMap` pointer with ownership
```
> `ptr` must be a non-null,

This is no longer a SAFETY precondition, but a PANICS precondition

apljungquist reviewed (commented):
Some nice improvements in this iteration. But there are still old comments that remain unaddressed. You may want to also take at least a quick look at Gustaf's comments as he he the person who will ultimately merge your PR if approved 😉

vsem-azamat committed:
fix(vdo): Fix soundness hole in file_descriptor() API

Return RawFd instead of BorrowedFd and mark the method unsafe.
BorrowedFd allowed callers to call try_clone_to_owned() in safe code,
which could lead to fd aliasing and use-after-close when VDO unrefs
the buffer.

vsem-azamat commented on crates/vdo/src/lib.rs:
```diff
@@ -461,54 +467,65 @@ impl StreamBuffer<'_> {
     /// Returns a borrowed file descriptor for the buffer's backing memory.
+    ///
+    /// The fd is only valid for the lifetime of this buffer. Do not convert it to
+    /// an `OwnedFd` (e.g. via `try_clone_to_owned`), as the underlying fd is closed
```
Good catch. Any safe API that allows users to trigger undefined behavior is unsound and must be marked unsafe.
Changed the return type from `BorrowedFd<'_>` to `RawFd` and marked the method `unsafe`.

vsem-azamat committed:
docs(vdo): Separate Safety and Panics sections in unsafe fn docs

vsem-azamat committed:
fix(vdo): Rewrite Send safety comments without GLib assumption

vsem-azamat committed:
refactor(vdo): Remove redundant RunningStream::stop() method

drop() achieves the same result and is idiomatic Rust.

vsem-azamat committed:
refactor(vdo): Use NonNull::new instead of assert + new_unchecked

vsem-azamat commented on crates/vdo/src/lib.rs:
```diff
@@ -0,0 +1,1359 @@
+                .to_string()
+        };
+
+        // Free the GError
```
Done - added `// SAFETY:` comments on unsafe blocks, e.g. [here](https://github.com/vsem-azamat/acap-rs/blob/feature/vdo-completion/crates/vdo/src/lib.rs#L110-L112).

vsem-azamat committed:
fix(vdo): Match error codes as i32 to avoid wrapping negative GError codes

vsem-azamat committed:
fix(vdo): Use == GFALSE for gboolean checks instead of != GTRUE

GLib treats any nonzero gboolean as true, not just GTRUE (1).
Using != GTRUE could misclassify a valid success value like 2 as failure.

vsem-azamat committed:
fix(vdo): Track stream started state to avoid stopping unstarted streams

Previously Drop always called vdo_stream_stop, even on streams that were
never started or where start() failed. This relied on undocumented VDO
idempotency. Now a `started` flag ensures stop is only called on streams
that were actually started successfully.

vsem-azamat committed:
fix(vdo): Fix buffer leak in StreamBuffer::unref() on error

mem::forget(self) was called before the FFI unref call, so on failure
the buffer was permanently leaked with Drop already suppressed. Move
mem::forget after the success check so Drop can retry cleanup on error.

vsem-azamat committed:
fix(vdo): Log StreamBuffer Drop failures even without GError

Previously Drop only logged when a GError was present, silently ignoring
GFALSE returns without an error object. Now logs in both cases.

vsem-azamat committed:
refactor(vdo): Minor code review cleanups

- Remove dead zero-arg arm from try_func! macro
- Use != GFALSE instead of != 0 in is_last_buffer for consistency
- Remove public Default derive from VdoError (code=0 misleadingly
  maps to VDO_ERROR_UNKNOWN; only used internally in from_gerror)
- Add production warning to Map::dump() doc comment

vsem-azamat committed:
fix(vdo): Use correct types for VdoFormat and fix Send safety comments

VdoFormat wraps c_int (i32) but was stored via set_u32, which would
wrap VDO_FORMAT_NONE (-1) to u32::MAX. Added Map::set_i32/get_i32
and switched the builder to use set_i32 for format.

Reworded Send safety comments - the justification is exclusive
ownership and no thread-pinning requirement, not absence of Sync.

vsem-azamat committed:
fix(vdo): Revert format to set_u32 - VDO API expects uint32

VdoFormat is c_int in bindgen, but VDO actually expects uint32
for the format key. Using set_i32 causes VDO to ignore the value.

apljungquist committed:
f: consistent cfg

https://github.com/AxisCommunications/acap-rs/pull/223#discussion_r2659585648

apljungquist committed:
Proper tests module and cfg

apljungquist committed:
update checksums

apljungquist committed:
fix `make check_docs`

apljungquist committed:
don't attempt to run vdo tests on host

guoxe reviewed (commented):
I have some questions relating to some of the functions added.
Overall I think it looks good!

vsem-azamat committed:
Merge branch 'main' into feature/vdo-completion

vsem-azamat committed:
address review comments

vsem-azamat committed:
update checksums

vsem-azamat committed:
document fd lifetime assumption

vsem-azamat commented on crates/vdo/src/map.rs:
```diff
@@ -0,0 +1,170 @@
+impl Map {
+    pub fn try_new() -> std::result::Result<Self, super::Error> {
+        let map = unsafe { vdo_sys::vdo_map_new() };
+        if map.is_null() {
```
Right, the null path was unreachable. 

`Map::new()` is now infallible and returns `Self` directly. 
Fixed in 7489918.

vsem-azamat commented on crates/vdo/src/map.rs:
```diff
@@ -0,0 +1,170 @@
+    }
+
+    pub fn set_bool(&mut self, key: &CStr, value: bool) {
+        let gvalue = if value {
```
Done: `set_bool/get_bool` use `IntoGlib/from_glib` from the glib crate. 

Fixed in 7489918.

vsem-azamat commented on crates/vdo/src/map.rs:
```diff
@@ -0,0 +1,170 @@
+    }
+
+    pub fn get_bool(&self, key: &CStr, default: bool) -> bool {
+        let gdefault = if default {
```
Same fix, see above

vsem-azamat commented on crates/vdo/src/lib.rs:
```diff
@@ -0,0 +1,1330 @@
+    /// Explicitly unreferences this buffer, returning an error if the operation fails.
+    ///
+    /// Normally buffers are unreferenced on drop. Use this if you need error handling.
+    pub fn unref(self) -> std::result::Result<(), Error> {
```
Removed, Drop handles cleanup: 7489918

vsem-azamat commented on crates/vdo/src/lib.rs:
```diff
@@ -0,0 +1,1330 @@
+unsafe impl Send for Stream {}
+
+impl Stream {
+    pub fn builder() -> StreamBuilder {
```
Removed; `StreamBuilder::new()` is now the only entry point.

vsem-azamat commented on crates/vdo/src/lib.rs:
```diff
@@ -451,426 +301,350 @@ impl StreamBuilder {
-    /// Returns the frame data as a mutable byte slice.
-    pub fn as_mut_slice(&mut self) -> Result<&mut [u8]> {
-        let data = unsafe { vdo_buffer_get_data(self.raw) };
+    /// Returns a copy of exactly [`size()`](StreamBuffer::size) bytes of frame data.
```
Fixed: now it offsets by `header_size()` and the doc comment states that the header is excluded. `as_slice()` remains for a raw view of the whole buffer.

vsem-azamat committed:
remove file_descriptor until there is a consumer

vsem-azamat committed:
update checksums

vsem-azamat commented on crates/vdo/src/lib.rs:
```diff
@@ -451,426 +301,350 @@ impl StreamBuilder {
-            std::os::fd::BorrowedFd::borrow_raw(fd)
+    /// The returned fd is owned by VDO and will be closed when this buffer is
+    /// unreferenced. The caller must not close or duplicate (`dup`) the fd.
+    pub unsafe fn file_descriptor(&self) -> std::result::Result<std::os::fd::RawFd, Error> {
```
I've removed `file_descriptor`. You're right, I'll add it when I test it (most likely in #230 ) and understand if it's really needed. 

Removed in 17f0d26.

guoxe commented on crates/vdo/src/lib.rs:
```diff
@@ -0,0 +1,1359 @@
+
+    /// Returns the file descriptor for the frame data.
+    ///
+    /// This can be used for zero-copy operations with other APIs.
```
Acknowledged

guoxe reviewed (dismissed):
Nice work!

Sorry it took a while for me to get around to reviewing, unless there is something else you want to address I am happy with merging this PR as is.

vsem-azamat committed:
update checksums
